In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None) 

In [2]:
df_pair_group = pd.read_csv('archivos/world_data_full.csv')

## EDA

In [3]:
df_pair_group.head()

Unnamed: 0      country density abbreviation  agricultural_land  land_area  \
0           0  Afghanistan      60           AF               58.1    652,230   
1           1      Albania     105           AL               43.1     28,748   
2           2      Algeria      18           DZ               17.4  2,381,741   
3           3      Andorra     164           AD               40.0        468   
4           4       Angola      26           AO               47.5  1,246,700   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   
3               NaN        7.20         376.0   Andorra la Vella   
4           117,000       40.73         244.0             Luanda   

  co2-emissions     cpi  cpi_change currency-code  fertility_rate  \
0         8,672   149.9         2.3           AFN            4.47   
1         4,536  119.05         1.4           ALL            1.62   
2       150,006  151.36         2.0           DZD            3.02   
3           469     NaN         NaN           EUR            1.27   
4        34,693  261.73        17.1           AOA            5.52   

   forested_area  gasoline_price               gdp  \
0            2.1            0.70   19,101,353,833    
1           28.1            1.36   15,278,077,447    
2            0.8            0.28  169,988,236,398    
3           34.0            1.51    3,154,057,987    
4           46.3            0.97   94,635,415,870    

   gross_primary_education_enrollment  gross_tertiary_education_enrollment  \
0                               104.0                                  9.7   
1                               107.0                                 55.0   
2                               109.9                                 51.4   
3                               106.4                                  NaN   
4                               113.5                                  9.3   

   infant_mortality      largest_city  life_expectancy  \
0              47.9             Kabul             64.5   
1               7.8            Tirana             78.5   
2              20.1           Algiers             76.7   
3               2.7  Andorra la Vella              NaN   
4              51.6            Luanda             60.8   

   maternal_mortality_ratio  minimum_wage official_language  \
0                     638.0          0.43            Pashto   
1                      15.0          1.12          Albanian   
2                     112.0          0.95            Arabic   
3                       NaN          6.63           Catalan   
4                     241.0          0.71        Portuguese   

   out_of_pocket_health_expenditure  physicians_per_thousand  population  \
0                              78.4                     0.28  38,041,754   
1                              56.9                     1.20   2,854,191   
2                              28.1                     1.72  43,053,054   
3                              36.4                     3.33      77,142   
4                              33.4                     0.21  31,825,295   

   population_labor_force_participation  tax_revenue  total_tax_rate  \
0                                  48.9          9.3            71.4   
1                                  55.7         18.6            36.6   
2                                  41.2         37.2            66.1   
3                                   NaN          NaN             NaN   
4                                  77.5          9.2            49.1   

   unemployment_rate urban_population     latitude   longitude  
0              11.12        9,797,273    33.93911    67.709953  
1              12.33        1,747,593   41.153332    20.168331  
2              11.70       31,510,100   28.033886     1.659626  
3                NaN

In [4]:
df_pair_group.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 36 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Unnamed: 0                            195 non-null    int64  
 1   country                               195 non-null    object 
 2   density                               195 non-null    object 
 3   abbreviation                          188 non-null    object 
 4   agricultural_land                     188 non-null    float64
 5   land_area                             194 non-null    object 
 6   armed_forces_size                     171 non-null    object 
 7   birth_rate                            189 non-null    float64
 8   calling_code                          194 non-null    float64
 9   capital/major_city                    192 non-null    object 
 10  co2-emissions                         188 non-null    object 
 11  cpi                

In [5]:
df_pair_group.describe(include='O').T

count unique              top freq
country              195    195      Afghanistan    1
density              195    137               25    7
abbreviation         188    188               AF    1
land_area            194    194          652,230    1
armed_forces_size    171    105            2,000    7
capital/major_city   192    192            Kabul    1
co2-emissions        188    184              143    2
cpi                  178    175           110.62    2
currency-code        180    133              EUR   23
gdp                  193    193  19,101,353,833     1
largest_city         189    188            S����    2
official_language    190     76          English   31
population           194    194       38,041,754    1
urban_population     190    190        9,797,273    1
latitude             195    195        33.93911     1
longitude            195    195        67.709953    1

In [6]:
df_pair_group.isnull().sum()

Unnamed: 0                               0
country                                  0
density                                  0
abbreviation                             7
agricultural_land                        7
land_area                                1
armed_forces_size                       24
birth_rate                               6
calling_code                             1
capital/major_city                       3
co2-emissions                            7
cpi                                     17
cpi_change                              16
currency-code                           15
fertility_rate                           7
forested_area                            7
gasoline_price                          20
gdp                                      2
gross_primary_education_enrollment       7
gross_tertiary_education_enrollment     12
infant_mortality                         6
largest_city                             6
life_expectancy                          8
maternal_mo

## Ejercicios apply

#### Las columnas land_area, armed_forces_size, co2-emissions, gdp, population, urban_population, latitude, longitude, density, cpi son de tipo object pero en realidad deberían ser numéricas. Crea una función que nos permita cambiar el tipo de dato de estas columnas al que le corresponde.

In [7]:
columnas= ['land_area', 'armed_forces_size', 'co2-emissions', 'gdp', 'population', 'urban_population', 'latitude', 'longitude', 'density', 'cpi']

In [8]:
df_pair_group.head(3)

Unnamed: 0      country density abbreviation  agricultural_land  land_area  \
0           0  Afghanistan      60           AF               58.1    652,230   
1           1      Albania     105           AL               43.1     28,748   
2           2      Algeria      18           DZ               17.4  2,381,741   

  armed_forces_size  birth_rate  calling_code capital/major_city  \
0           323,000       32.49          93.0              Kabul   
1             9,000       11.78         355.0             Tirana   
2           317,000       24.28         213.0            Algiers   

  co2-emissions     cpi  cpi_change currency-code  fertility_rate  \
0         8,672   149.9         2.3           AFN            4.47   
1         4,536  119.05         1.4           ALL            1.62   
2       150,006  151.36         2.0           DZD            3.02   

   forested_area  gasoline_price               gdp  \
0            2.1            0.70   19,101,353,833    
1           28.1            1.36   15,278,077,447    
2            0.8            0.28  169,988,236,398    

   gross_primary_education_enrollment  gross_tertiary_education_enrollment  \
0                               104.0                                  9.7   
1                               107.0                                 55.0   
2                               109.9                                 51.4   

   infant_mortality largest_city  life_expectancy  maternal_mortality_ratio  \
0              47.9        Kabul             64.5                     638.0   
1               7.8       Tirana             78.5                      15.0   
2              20.1      Algiers             76.7                     112.0   

   minimum_wage official_language  out_of_pocket_health_expenditure  \
0          0.43            Pashto                              78.4   
1          1.12          Albanian                              56.9   
2          0.95            Arabic                              28.1   

   physicians_per_thousand  population  population_labor_force_participation  \
0                     0.28  38,041,754                                  48.9   
1                     1.20   2,854,191                                  55.7   
2                     1.72  43,053,054                                  41.2   

   tax_revenue  total_tax_rate  unemployment_rate urban_population  \
0          9.3            71.4              11.12        9,797,273   
1         18.6            36.6              12.33        1,747,593   
2         37.2            66.1              11.70       31,510,100   

     latitude   longitude  
0   33.93911    67.709953  
1  41.153332    20.168331  
2  28.033886     1.659626

Para generar la función mirar por columnas

In [ ]:
def numerica(columna):

for columna in columnas:
        columna = columna.replace(',','')
        columna = columna.dtype(float)

return(columna)

ValueError: could not convert string to float: '2.381.741'

In [14]:
df_pair_group['land_area']= df_pair_group['land_area'].apply(numerica)

AttributeError: 'str' object has no attribute 'dtype'

In [ ]:
df_pair_group['land_area'].isnull().sum()

30

In [ ]:
df_pair_group['land_area'].value_counts(dropna=False)

land_area
NaN        30
652.230     1
312.685     1
323.802     1
309.500     1
           ..
245.857     1
36.125      1
214.969     1
27.750      1
390.757     1
Name: count, Length: 166, dtype: int64

Investigamos los nulos iniciales y cuántos nulos hemos generado con el except para valorar opciones. En total hemos generado 29 nulos de 195 datos.

#### Crea una función para añadir una nueva columna llamada 'Continent' al DataFrame para indicar a qué continente pertenece cada país

chat gpt: generar un diccionario clave/valor, continente/país pasándole la lista que tenemos en el DataFrame

In [ ]:
continentes = {
    'Asia': ["Philippines", 'Afghanistan', 'Armenia', 'Azerbaijan', 'Bahrain', 'Bangladesh', 'Bhutan', 'Brunei', 'Cambodia', 'China', 'East Timor', 'India', 'Indonesia', 'Iran', 'Iraq', 'Israel', 'Japan', 'Jordan', 'Kazakhstan', 'Kuwait', 'Kyrgyzstan', 'Laos', 'Lebanon', 'Malaysia', 'Maldives', 'Mongolia', 'Myanmar', 'Nepal', 'North Korea', 'Oman', 'Pakistan', 'Palestinian National Authority', 'Qatar', 'Saudi Arabia', 'Singapore', 'South Korea', 'Sri Lanka', 'Syria', 'Tajikistan', 'Thailand', 'Turkmenistan', 'United Arab Emirates', 'Uzbekistan', 'Vietnam', 'Yemen'],
    'Africa': ["Tunisia", "Ivory Coast", 'Algeria', 'Angola', 'Benin', 'Botswana', 'Burkina Faso', 'Burundi', 'Cape Verde', 'Cameroon', 'Central African Republic', 'Chad', 'Comoros', 'Republic of the Congo', 'Cote d\'Ivoire', 'Democratic Republic of the Congo', 'Djibouti', 'Egypt', 'Equatorial Guinea', 'Eritrea', 'Eswatini', 'Ethiopia', 'Gabon', 'The Gambia', 'Ghana', 'Guinea', 'Guinea-Bissau', 'Kenya', 'Lesotho', 'Liberia', 'Libya', 'Madagascar', 'Malawi', 'Mali', 'Mauritania', 'Mauritius', 'Morocco', 'Mozambique', 'Namibia', 'Niger', 'Nigeria', 'Rwanda', 'Sao Tome and Principe', 'Senegal', 'Seychelles', 'Sierra Leone', 'Somalia', 'South Africa', 'South Sudan', 'Sudan', 'Tanzania', 'Togo', 'Uganda', 'Zambia', 'Zimbabwe'],
    'Europe': ["Turkey", "Republic of Ireland", "Vatican City", 'Albania', 'Andorra', 'Austria', 'Belarus', 'Belgium', 'Bosnia and Herzegovina', 'Bulgaria', 'Croatia', 'Cyprus', 'Czech Republic', 'Denmark', 'Estonia', 'Finland', 'France', 'Georgia', 'Germany', 'Greece', 'Hungary', 'Iceland', 'Ireland', 'Italy', 'Latvia', 'Liechtenstein', 'Lithuania', 'Luxembourg', 'Malta', 'Moldova', 'Monaco', 'Montenegro', 'Netherlands', 'North Macedonia', 'Norway', 'Poland', 'Portugal', 'Romania', 'Russia', 'San Marino', 'Serbia', 'Slovakia', 'Slovenia', 'Spain', 'Sweden', 'Switzerland', 'Ukraine', 'United Kingdom'],
    "Central America":["Belize", "Saint Vincent and the Grenadines", "Saint Lucia", "Saint Kitts and Nevis", "Panama", "Nicaragua", "Mexico", "Jamaica", "Honduras", "Haiti", "Guatemala", "Grenada", "El Salvador", "Dominican Republic", "Dominica", "Cuba", "Antigua and Barbuda", "The Bahamas", "Barbados"," Belize", "Costa Rica",  ], 
    'North America': ['Canada', 'United States'],
    'South America': ["Trinidad and Tobago", 'Argentina', 'Bolivia', 'Brazil', 'Chile', 'Colombia', 'Ecuador', 'Guyana', 'Paraguay', 'Peru', 'Suriname', 'Uruguay', 'Venezuela'],
    'Oceania': ["Papua New Guinea", "Federated States of Micronesia", 'Australia', 'Fiji', 'Kiribati', 'Marshall Islands', 'Nauru', 'New Zealand', 'Palau', 'Samoa', 'Solomon Islands', 'Tonga', 'Tuvalu', 'Vanuatu']
}

#### Guarda el DataFrame después de los cambios creados después de los dos ejercicios anteriores para usarlo posteriormente.